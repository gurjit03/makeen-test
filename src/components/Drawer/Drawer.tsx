import React from 'react';

const Drawer = () => {
    return (
        <div>DRAWER</div>
    )
}

export default Drawer;