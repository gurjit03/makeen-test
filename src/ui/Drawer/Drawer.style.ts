import styled, {css} from 'styled-components'
import { colors } from '../../config/jss';

const transitionSpeed = 0.7

export const DrawerWrapper = styled.div<{isOpen: boolean}>`
transition: transform ${transitionSpeed}s ease;
    background-color: ${colors.white};
    max-width: 400px;
    width: 100%;
    height: 100%;
    transform: translateX(100%);
    z-index: 1000;
    overflow: auto;
    position: fixed;
    box-shadow: 0 0 15px rgba(0, 0, 0, 0.5);
    transition: transform ${transitionSpeed}s ease;
    ${({isOpen}) => isOpen && css`
        top: 0;
        right: 0;
        transform: translateX(0);
    `}
`;

export const Backdrop = styled.div<{isOpen: boolean}>`
    visibility: hidden;
    opacity: 0;
    background: rgba(0, 0, 0, 0.5);
    transition: opacity ${transitionSpeed}s ease,
        visibility ${transitionSpeed}s ease;
    width: 100vw;
    height: 100vh;
    top: 0;
    right: 0;
    position: fixed;
    pointer-events: none;
	z-index: 0;
    ${({isOpen}) => isOpen && css`
        visibility: visible;
        opacity: 1;
        pointer-events: auto;
        z-index: 999;
    `}
`;
