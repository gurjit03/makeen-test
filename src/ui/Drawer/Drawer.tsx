import React from 'react'
import { Backdrop, DrawerWrapper } from './Drawer.style'

interface DrawerProps {
	isOpen: boolean
	onClose?: () => void
	children: React.ReactChild
}

const Drawer: React.FC<DrawerProps> = props => {
	const { isOpen, onClose = () => {}, children } = props
	const bodyRef = React.useRef<HTMLElement>(document.querySelector('body'))

	React.useEffect(() => {
		const bodyEl = bodyRef.current

		const updatePageScroll = () => {
			if (bodyEl) {
				if (isOpen) {
					bodyEl.style.overflow = 'hidden'
				} else {
					bodyEl.style.overflow = ''
				}
			}
		}
		updatePageScroll()

		return () => {
			if (bodyEl) {
				bodyEl.style.overflow = ''
			}
		}
	}, [isOpen])

	return (
		<div>
			<DrawerWrapper isOpen={isOpen}>{children}</DrawerWrapper>
			<Backdrop role='button' isOpen={isOpen} onClick={onClose} />
		</div>
	)
}

export default Drawer
