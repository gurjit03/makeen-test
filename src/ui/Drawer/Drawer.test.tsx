import React from 'react';
import { render, fireEvent, screen, waitFor } from '@testing-library/react';
import Drawer from './Drawer'


describe('<Drawer />', () => {
    it('should render properly', () => {
        const { asFragment } = render(<Drawer isOpen>content</Drawer>)
        expect(asFragment()).toMatchSnapshot();
    });

    it('should display the content properly', () => {
        render(<Drawer isOpen>content</Drawer>)
        expect(screen.getByText('content')).toBeInTheDocument()
    });

    it('should causes body element to have styles [overflow="hidden"] on isOpen="true" and [overflow="auto"] on isOpen="false"', async () => {
        const { rerender } = render(<Drawer isOpen>content</Drawer>)
        expect(document.body).toHaveStyle({
            overflow: 'hidden'
        })

        rerender(<Drawer isOpen={false}>content</Drawer>)
        await waitFor(() => {
            expect(document.body).not.toHaveStyle({
                overflow: 'hidden'
            })
        });
    })

    it('should call onClose handler on backdrop click', () => {
        const onClickHandler = jest.fn();
        render(<Drawer isOpen onClose={onClickHandler}>content</Drawer>)
        const backdropEl = screen.getByRole('button')
        fireEvent.click(backdropEl)
        expect(onClickHandler).toBeCalledTimes(1);
    });
})