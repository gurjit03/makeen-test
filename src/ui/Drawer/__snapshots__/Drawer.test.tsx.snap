// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Drawer /> should render properly 1`] = `
<DocumentFragment>
  <div>
    <div
      class="sc-bdfBQB Pynap"
    >
      content
    </div>
    <div
      aria-pressed="false"
      class="sc-gsTEea bMysJp"
      role="button"
    />
  </div>
</DocumentFragment>
`;
